presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -991,6 +1008,52 @@ private Tuple2<MutablePartitionId, T> doComputeNext()
             throw new RuntimeException(failure);
         }
 
+        private void collectMetrics(TaskInfo taskInfo)
+        {
+            int taskId = taskInfo.getTaskId().getId();
+            int stageId = taskInfo.getTaskId().getStageExecutionId().getStageId().getId();
+            List<Map<String, String>> newStatsList = new ArrayList<>();
+            for (PipelineStats pipelineStats : taskInfo.getStats().getPipelines()) {
+                for (OperatorStats operatorStats : pipelineStats.getOperatorSummaries()) {
+                    if (operatorStats.getOperatorType().equals("NativeExecutionOperator")) {
+                        NativeExecutionInfo nativeExecutionInfo = (NativeExecutionInfo) operatorStats.getInfo();
+                        for (TaskStats taskStats : nativeExecutionInfo.getTaskStats()) {
+                            RuntimeStats runtimeStat = taskStats.getRuntimeStats();
+                            Map<String, String> newStatMap = new HashMap<>();
+                            for (Map.Entry<String, RuntimeMetric> entry : runtimeStat.getMetrics().entrySet()) {
+                                String key = entry.getKey();
+                                RuntimeMetric metric = entry.getValue();
+                                if (metric.getCount() == 0) {
+                                    continue;
+                                }
+                                String metricSumKey = buildGenericMetricKey(taskId, stageId, key, "sum");
+                                newStatMap.put(metricSumKey, String.valueOf(metric.getSum()));
+                                String metricCountKey = buildGenericMetricKey(taskId, stageId, key, "count");
+                                newStatMap.put(metricCountKey, String.valueOf(metric.getCount()));
+                                String metricMinKey = buildGenericMetricKey(taskId, stageId, key, "min");
+                                newStatMap.put(metricMinKey, String.valueOf(metric.getMin()));
+                                String metricMaxKey = buildGenericMetricKey(taskId, stageId, key, "max");
+                                newStatMap.put(metricMaxKey, String.valueOf(metric.getMax()));
+                                String metricUnitKey = buildGenericMetricKey(taskId, stageId, key, "unit");
+                                newStatMap.put(metricUnitKey, String.valueOf(metric.getUnit()));
+                            }
+                            if (!newStatMap.isEmpty()) {
+                                newStatsList.add(newStatMap);
+                            }
+                        }
```

Comment:
This is too deeply nested, I would suggest moving this to function or re-writing using streams API and lambda.

```
Map<String, Function<RuntimeMetric, Object>> metricFunctions = new HashMap<>();
            metricFunctions.put("sum", RuntimeMetric::getSum);
            metricFunctions.put("count", RuntimeMetric::getCount);
            metricFunctions.put("min", RuntimeMetric::getMin);
            metricFunctions.put("max", RuntimeMetric::getMax);
            metricFunctions.put("unit", RuntimeMetric::getUnit);

            List<Map<String, String>> newStatsList = new ArrayList<>();
            taskInfo.getStats().getPipelines().stream()
                    .flatMap(pipelineStats -> pipelineStats.getOperatorSummaries().stream())
                    .filter(operatorStats -> operatorStats.getOperatorType().equals("NativeExecutionOperator"))
                    .map(operatorStats -> (NativeExecutionInfo) operatorStats.getInfo())
                    .flatMap(nativeExecutionInfo -> nativeExecutionInfo.getTaskStats().stream())
                    .forEach(taskStats -> {
                        RuntimeStats runtimeStats = taskStats.getRuntimeStats();
                        Map<String, String> newStatMap = new HashMap<>();
                        runtimeStats.getMetrics().entrySet().stream()
                                .filter(entry -> entry.getValue().getCount() > 0)
                                .forEach(entry -> {
                                    String key = entry.getKey();
                                    RuntimeMetric metric = entry.getValue();
                                    metricFunctions.forEach((metricFunctionKey, metricFunction) -> {
                                        String metricKey = buildGenericMetricKey(taskId, stageId, key, metricFunctionKey);
                                        newStatMap.put(metricKey, String.valueOf(metricFunction.apply(metric)));
                                    });
                                });
                        if (!newStatMap.isEmpty()) {
                            newStatsList.add(newStatMap);
                        }
                    });
            if (!newStatsList.isEmpty()) {
                genericShuffleStatsCollector.add(newStatsList);
            }
```

Reply:
+1